Add Mesos Installer

mattboehm opened the pull request:
Running `ansible-playbook provision-mesos.yml` will install onto mesos

* Read many vars from the env
* Don't use sudo
* Only run ckan/ace tasks
* Run apache in the foreground
* Store all files in MESOS_SANDBOX (and store ckan user uploads on the network drive)

mattboehm committed:
Add support for installing onto aurora/mesos

Running `ansible-playbook provision-mesos.yml` will install onto mesos

* Read many vars from the env
* Don't use sudo
* Only run ckan/ace tasks
* Run apache in the foreground
* Store all files in MESOS_SANDBOX (and store ckan user uploads on the network drive)

mattboehm committed:
Add instructions for running on marathon

mattboehm committed:
Change apache port and fix permission issue

mattboehm commented:
Everything should be working now but before merging, I'd like to do a complete vagrant destroy/reinstall for both the traditional and mesos installations to ensure that nothing broke. If anyone else would like to try the mesos installer, I've added instructions to the readme.

tanderegg commented on README.md:
```diff
@@ -27,6 +27,22 @@ Installation instructions can be found here: http://docs.vagrantup.com/v2/instal
+6. Edit the app's config, click the "json" toggle in the upper right, and paste the json from `marathon_app.json` (located in the same directory as this readme). Update any instances of `10.0.2.2` with the gateway IP you noted in step 4.
+7. Upon saving the config, you'll be prompted to (re)deploy the application. Do so and give it time to install (~10 minutes?)
+8. Download the stdout/stderr files in marathon and briefly scan them for errors
+9. Within those stdout/stderr, note the long directory that everything is being put into (`MESOS_SANDBOX`). Try `vagrant ssh mesos_agent_1 -- -L 8888:localhost:8888` and within that machine `ls <path/to/sandbox>`. If the file does not exist, kill the ssh and retry with `mesos_agent_2` and `mesos_agent_3` until you've found the agent that is running the application.
```
In the Marathon UI, if you go to the app and "instances", it will list the running instances and the IP address they are deployed to, which is probably the easiest way to see which Agent it is running on.

tanderegg commented on README.md:
```diff
@@ -27,6 +27,22 @@ Installation instructions can be found here: http://docs.vagrantup.com/v2/instal
+9. Within those stdout/stderr, note the long directory that everything is being put into (`MESOS_SANDBOX`). Try `vagrant ssh mesos_agent_1 -- -L 8888:localhost:8888` and within that machine `ls <path/to/sandbox>`. If the file does not exist, kill the ssh and retry with `mesos_agent_2` and `mesos_agent_3` until you've found the agent that is running the application.
+10. Try visiting [localhost:8888](http://localhost:8888) and ensure that you can see the Data Catalog page.
+11. If you encounter an apache error page, go to your ssh session with the mesos agent and look for details in the logs at `/path/to/sandbox/httpd`
+
```
Note that you can also view the sandbox through the Mesos UI, if you go to 10.0.1.31:5050 for example, click on the "sandbox" link next to the running task, and you can browse it in the browser.

tanderegg commented on marathon_app.json:
```diff
@@ -0,0 +1,49 @@
+  "cmd": "env; cd ckan-installer-mesos/provisioners && ANSIBLE_CONFIG=ansible-mesos.cfg ansible-playbook -vvvv provision-mesos.yml ; httpd -DFOREGROUND -f \"$MESOS_SANDBOX/apache_ckan.conf\"; sleep 600",
+  "cpus": 1,
+  "mem": 100,
+  "disk": 800,
```
Normally I try to use RAM as a base 2 value, i.e. 128 instead of 100, but this is not strictly necessary.  Also, we don't really use disk isolation yet, so setting "disk" to 0 is fine for now.

tanderegg commented on marathon_app.json:
```diff
@@ -0,0 +1,49 @@
+{
+  "id": "/dc-test",
+  "cmd": "env; cd ckan-installer-mesos/provisioners && ANSIBLE_CONFIG=ansible-mesos.cfg ansible-playbook -vvvv provision-mesos.yml ; httpd -DFOREGROUND -f \"$MESOS_SANDBOX/apache_ckan.conf\"; sleep 600",
```
It might be cleaner just to put ANSIBLE_CONFIG in the "env" block below rather in the command itself.

tanderegg commented on marathon_app.json:
```diff
@@ -0,0 +1,49 @@
+  },
+  "healthChecks": [
+    {
+      "path": ":8888/",
```
Having the port here explicitly shouldn't be needed, the "portIndex" 0 below will explicitly use the port assigned to this instance of the app.  Is Apache setup to use $PORT0?

tanderegg commented on marathon_app.json:
```diff
@@ -0,0 +1,49 @@
+  "uris": [
+    "https://github.com/mattboehm/ckan-installer/archive/mesos.tar.gz"
+  ],
+  "fetch": [
```
The "fetch" block isn't strictly needed, as it is autogenerated by Marathon from the "uris" block.  I usuall remove it from the JSON definition, as things like "cache" are dynamically set depending on how Marathon/Mesos are configured.

mattboehm commented:
I've tested both forms of installations and am having Valeri run through the instructions now to try it for himself.

mattboehm committed:
Update mesos installation instructions

Fix a few issues discovered by Valeri

mattboehm committed:
Change file upload storage dir to /home/work

mattboehm committed:
Update default site_url to support newer versions of ckan

mattboehm committed:
Create apache run dir

needed for pid file.

tanderegg commented on provisioners/group_vars/mesos:
```diff
@@ -0,0 +1,45 @@
+apache_as_service: False
+apache_config_dir: "{{lookup('env', 'MESOS_SANDBOX')}}/apache_ckan.conf"
+apache_server_root: "{{lookup('env', 'MESOS_SANDBOX')}}/httpd"
+apache_port: 8888
```
This should probably be {{ lookup('env', 'PORT0') }}, to pull in the port from Marathon, no?

mattboehm committed:
Recurse when creating ckan directories

mattboehm committed:
Mesos: use dynamic port, adjust resources, simplify docs

This should address all of Tim's PR feedback and simplify the
installation process

mattboehm committed:
Add some documentation on configuring/upgrading to the readme

mattboehm committed:
Add more configurable variables for mesos-staging, fix typos

mattboehm committed:
Fix ckan site_url for mesos

mattboehm committed:
make ckan user/group configurable

mattboehm committed:
only add run dir if it doesn't exist

mattboehm committed:
correctly test if ckan_ldap_uri is defined

mattboehm committed:
Make ckan user/group default to mesagent if currently falsy

mattboehm committed:
Fix ckan file storage dir default

mattboehm committed:
Comment out ownership of upload dir for now

mattboehm committed:
Urlencode database passwords

mattboehm committed:
Try fixing yaml comment

mattboehm committed:
Undo urlencoding

Should be unnecessary as it wasn't needed in prod instance

mattboehm committed:
Change header for single sign on

mattboehm committed:
Merge remote-tracking branch 'matt/mesos-rebase' into mesos-rebase

mattboehm committed:
Only create upload dir if it doesn't exist

mattboehm committed:
Add export and sso plugins

mattboehm committed:
Fix single sign on header

mattboehm committed:
Add owner/group back into "create upload dir" task

mattboehm committed:
upgrade ckan tables and re-index solr instances

mattboehm committed:
Read ckan version from env var

mattboehm committed:
Make file modes more restrictive

mattboehm committed:
Use different file permissions on local

Since local vagrant installations run apache with a separate user, we
open the permissions up so that all users can read files. In production
on mesos, everything is done by 1 user so we can lock things down a bit
more.

tanderegg commented on marathon_app.json:
```diff
@@ -0,0 +1,48 @@
+  ],
+  "fetch": [
+    {
+      "uri": "https://github.com/mattboehm/ckan-installer/archive/mesos.tar.gz",
```
@mattboehm This should be changed to the primary repo before merging

tanderegg replied:
Ah, and actually this should be the "uris" key not the "fetch" key.  Basically, "fetch" is autogenerated by Marathon for each item in the "uris" lists.  So you just need:
```
"uris": {
  "https://github.com..."
}
```

tanderegg commented on provisioners/roles/ckan/tasks/main.yml:
```diff
@@ -20,38 +19,74 @@
-    # virtualenv is accessible.  
-    PATH: /usr/local/bin:/bin:/usr/bin
-  sudo: yes
+  shell: "source {{ckan_venv}}/bin/activate && python -m pip install -e git+https://github.com/ckan/ckan.git@{{ckan_version_tag}}#egg=ckan"
```
@mattboehm You could also just do `{{ckan_venv}}/bin/python -m ...` instead of activating the virtualenv

mattboehm replied:
I tend to prefer this since it updates a bunch of environment variables and I'm never quite sure if a pip install is going to shell out to some command that depends on those variables being set.

mattboehm committed:
Add import plugin

mattboehm committed:
Adjust permissions

mattboehm committed:
Add uwsgi as a dependency

tanderegg commented on provisioners/roles/ckan/tasks/main.yml:
```diff
@@ -20,38 +19,74 @@
-    # time gathering facts. We could just use anisble_env.PATH
-    PATH: /usr/pgsql-9.4/bin:/usr/local/bin:/bin:/usr/bin
-  sudo: yes
+  shell: "source {{ckan_venv}}/bin/activate && PATH=\"/usr/pgsql-9.4/bin:/usr/pgsql-9.5/bin:$PATH\" python -m pip install -r {{ckan_venv}}/src/ckan/requirements.txt"
```
Same here, you could explicitly use the virtualenv's python, and also potentially use Ansible's "environment" option like here: https://github.com/cfpb/aurora/blob/2ad6bd89dec788cc787f9b4883e82d0bf25ff25c/deploy/roles/sqitch/tasks/main.yml#L53, but with ansible_env.PATH to retrieve the target's PATH variable.

mattboehm replied:
ansible_env.PATH no longer works without further fiddling: https://github.com/ansible/ansible/issues/14655

The workarounds in that thread seem flaky and complicated, so for now I'm going to revert to setting the path inline but am open to suggestions.

tanderegg commented on provisioners/roles/ckan/tasks/main.yml:
```diff
@@ -60,105 +95,108 @@
-  sudo: yes
+  #sudo: yes
   #sudo_user: postgres
   #when: create_datastore_db|changed or create_datastore_db_user|changed or create_ckan_db|changed or create_ckan_db_user|changed
```
Should we remove the commented out lines at this point?

tanderegg commented:
@mattboehm Just some minor stuff, thanks!

mattboehm committed:
Try commenting out permissions

mattboehm committed:
Clean up code, address PR feedback

mattboehm committed:
Change path the old way

The fancier way breaks in newer versions of ansible

mattboehm committed:
Make plugin url configurable for mesos installations

mattboehm committed:
Disable sso for now

Need to resolve bug for unregistered users

mattboehm committed:
Add more ldap settings

mattboehm committed:
Add ckan ldap plugin

mattboehm committed:
Use Tim's fork of ckanext-ldap to support digest_md5

mattboehm committed:
Re-enable sso

mattboehm committed:
Don't create users via web on mesos installation

Instead, rely on ldap.

mattboehm committed:
Ignore errors if uploads dir can't be created

tanderegg commented:
@mattboehm Is the [WIP] label still applicable, or is this good to go from your end?  Looks correct to me, so I'm ready to merge if you are.

tanderegg commented:
@mattboehm bump

mattboehm commented:
@tanderegg whoops, I've removed the WIP label. We've been using this in prod for a while now, so it should be good to merge.

tanderegg commented:
@mattboehm Cool, that is what I thought, just wanted to be prudent :)
